File: 2019/day06/day06.c
#if 0
tmpfile=$(mktemp)
cc "$0" -o "$tmpfile" && "$tmpfile" "$@"
exit
#endif

#include <errno.h>
#include <stdio.h>
#include <inttypes.h>

#include "../c_utils/vec.h"
#include "../c_utils/map.h"

typedef struct {
    char label[4];
} orbit_label_t;

uint64_t orbit_label_t_hasher(const char* p) {
    const orbit_label_t* a = (const orbit_label_t*)p;
    return generic_hash(a->label, strlen(a->label));
}

int orbit_label_t_eq(const char* p, const char* q) {
    const orbit_label_t* a = (const orbit_label_t*)p;
    const orbit_label_t* b = (const orbit_label_t*)q;
    return strcmp(a->label, b->label) == 0;
}

typedef vec_t(orbit_label_t) vec_orbit_label_t;
typedef map_t(orbit_label_t, orbit_label_t) orbit_tree_t;

void path_from_root(orbit_tree_t* tree, const orbit_label_t* node,
                    vec_orbit_label_t* out) {
    for (;;) {
        map_get(tree, *node);
        if (tree->value == NULL) {
            break;
        }
        node = tree->value;
        vec_push(out, *node);
    }
    vec_reverse(out);
}

int32_t shortest_path_between(orbit_tree_t* tree, const orbit_label_t* start,
                              const orbit_label_t* end) {
    vec_orbit_label_t path1;
    vec_init(&path1);
    vec_orbit_label_t path2;
    vec_init(&path2);

    path_from_root(tree, start, &path1);
    path_from_root(tree, end, &path2);

    int32_t common_subpath_length = 0;
    for (size_t i = 0; ; i++) {
        if (i >= path1.length || i >= path2.length) {
            break;
        }

        if (strcmp(path1.data[i].label, path2.data[i].label) != 0) {
            break;
        }
        common_subpath_length++;
    }

    int32_t result = (int32_t)path1.length + (int32_t)path2.length
        - 2 * common_subpath_length;

    vec_done(&path2);
    vec_done(&path1);

    return result;
}

void read_input(char* filename, orbit_tree_t* out) {
    FILE* file = fopen(filename, "r");
    if (file == NULL) {
        printf("Error opening file %s: %d\n", filename, errno);
        exit(1);
    }

    orbit_label_t parent;
    orbit_label_t child;
    while (fscanf(file, "%3[^)\n])%3[^)\n] ", parent.label, child.label) == 2) {
        map_insert(out, child, parent);
    }

    fclose(file);
}

void star1(orbit_tree_t* tree) {
    vec_orbit_label_t path;
    vec_init(&path);

    size_t count = 0;
    map_foreach(tree) {
        vec_clear(&path);
        path_from_root(tree, tree->key, &path);
        count += path.length;
    }

    printf("%zu\n", count);

    vec_done(&path);
}

void star2(orbit_tree_t* tree) {
    orbit_label_t start = { "YOU" };
    orbit_label_t end = { "SAN" };

    printf("%" PRId32 "\n", shortest_path_between(tree, &start, &end));
}

int main(int argc, char** argv) {
    char* filename = "input.txt";
    if (argc > 1) {
        filename = argv[1];
    }

    orbit_tree_t tree;
    map_init(&tree, orbit_label_t_hasher, orbit_label_t_eq);

    read_input(filename, &tree);

    star1(&tree);
    star2(&tree);

    map_done(&tree);
}

File: 2019/day06/day06.rs
#!/usr/bin/env run-cargo-script
use std::collections::HashMap;

type DirectedTree = HashMap<String, String>;

fn path_to_root<'a>(tree: &'a DirectedTree, mut node: &'a str) -> Vec<&'a str> {
    let mut result = Vec::new();
    while tree.contains_key(node) {
        node = &tree[node];
        result.push(node);
    }
    result
}

fn shortest_path_between(tree: &DirectedTree, start: &str, end: &str) -> i32 {
    let mut path1 = path_to_root(tree, start);
    let mut path2 = path_to_root(tree, end);
    path1.reverse();
    path2.reverse();

    let mut common_subpath_length = 0;
    for (i, j) in path1.iter().zip(path2.iter()) {
        if i != j {
            break
        }
        common_subpath_length += 1;
    }

    (path1.len() + path2.len() - 2 * common_subpath_length) as i32
}

fn star1(orbits: &DirectedTree) -> i32 {
    let mut count = 0;
    for mut i in orbits.keys() {
        while orbits.contains_key(i) {
            i = &orbits[i];
            count += 1;
        }
    }

    count
}

fn star2(orbits: &DirectedTree) -> i32 {
    shortest_path_between(orbits, "YOU", "SAN")
}

fn main() {
    let args: Vec<_> = std::env::args().collect();
    let filename = args.get(1).map(|s| &s[..]).unwrap_or("input.txt");
    let input: DirectedTree = std::fs::read_to_string(filename)
        .unwrap()
        .trim()
        .split_whitespace()
        .map(|l| {
            let mut orbit = l.split(")");
            let parent = orbit.next().unwrap();
            let child = orbit.next().unwrap();
            (child.to_owned(), parent.to_owned())
        })
        .collect();

    println!("{}", star1(&input));
    println!("{}", star2(&input));
}
